### PR TITLE
Fix re connection

### PR DESCRIPTION
**Pull request by ashcherbakov**

### CODE REVIEW

File: stp_zmq/zstack.py [hunk 1]
```diff
@@ -1163,10 +1166,16 @@ def retryDisconnected(self, exclude=None):
...
-            else:
+
+            if not name in self._retry_connect:
+                self._retry_connect[name] = 0
+
+            if not remote.socket or self._retry_connect[name] >= KITZStack.MAX_RECONNECT_RETRY_ON_SAME_SOCKET:
+                self._retry_connect.pop(name, None)
                 self.reconnectRemote(remote)
+            else:
+                self._retry_connect[name] =  self._retry_connect.pop(name) + 1
```
Comment (lovesh) [inline stp_zmq/zstack.py]: Why not `self._retry_connect[name] += 1`?, why do a costly `pop`?

File: stp_zmq/test/test_reconnect.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    disconnected_motor = motors[0]
+    other_stacks = stacks[1:]
+
+    looper.removeProdable(disconnected_motor)
+    disconnected_motor.stop()
+
+    return disconnected_motor, other_stacks
+
+def test_reconnect_short(looper, connected_stacks, connection_timeout, disconnect_first_stack):
```
Comment (lovesh) [inline stp_zmq/test/test_reconnect.py]: Is the purpose of the test "to check that if a stack is kept disconnected for some time, it is able to reconnect"?

File: stp_zmq/test/test_reconnect.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        checkStacksConnected, other_stacks, retryWait=1, timeout=connection_timeout))
+
+    looper.runFor(1)
+
+    # CONNECT
+    looper.add(disconnected_motor)
+    looper.run(eventually(
+        checkStacksConnected, stacks, retryWait=1, timeout=connection_timeout))
+
+def test_reconnect_short(looper, connected_stacks, connection_timeout, disconnect_first_stack):
```
Comment (lovesh) [inline stp_zmq/test/test_reconnect.py]: Why the same name for test?

File: stp_zmq/test/test_reconnect.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    looper.run(eventually(
+        checkStackDisonnected, disconnected_motor.stack, other_stacks, retryWait=1, timeout=connection_timeout))
+    looper.run(eventually(
+        checkStacksConnected, other_stacks, retryWait=1, timeout=connection_timeout))
+
+    looper.runFor(1)
+
+    # CONNECT
+    looper.add(disconnected_motor)
+    looper.run(eventually(
```
Comment (lovesh) [inline stp_zmq/test/test_reconnect.py]: How is this test different from previous tests except using a longer timeout at last? All the conditions in both tests are same then why would the final check take twice the time?

File: stp_zmq/test/test_reconnect.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        checkStacksConnected, stacks, retryWait=1, timeout=2 * connection_timeout))
+
+
+def test_reconnect_long(looper, connected_stacks, connection_timeout, disconnect_first_stack):
```
Comment (lovesh) [inline stp_zmq/test/test_reconnect.py]: If the purpose of this test and 2 above is just to keep a stack disconnected for variable (in each test) duration of time, then why not abstract the lines before and after `looper.runFor` in 2 functions, why duplicate code?

File: stp_zmq/zstack.py [hunk 6]
```diff
@@ -1163,10 +1166,16 @@ def retryDisconnected(self, exclude=None):
         for name, remote in self.remotes.items():
             if name in exclude or remote.isConnected:
                 continue
-            if remote.socket:
-                self.sendPingPong(remote, is_ping=True)
-            else:
+
```
Comment (lovesh) [inline stp_zmq/zstack.py]: I am not seeing how this retry mechanism is tested by the above tests, this retry mechanism should be tested by patching one stack's ping pong handler to not send pong for any ping and see that retry attempts are counted and after `MAX_RECONNECT_RETRY_ON_SAME_SOCKET*RETRY_TIMEOUT_*` time, the other stack does a `reconnectRemote`, in another test patch should be reverted under `t` where `RETRY_TIMEOUT_*
 < t < MAX_RECONNECT_RETRY_ON_SAME_SOCKET*RETRY_TIMEOUT_`

Comment (ashcherbakov) [inline stp_zmq/zstack.py]: Added a test in https://github.com/evernym/stp/pull/33

File: stp_zmq/zstack.py (hunk 1 above)
Comment (ashcherbakov) [inline stp_zmq/zstack.py]: Fixed

File: stp_zmq/test/test_reconnect.py (hunk 2 above)
Comment (ashcherbakov) [inline stp_zmq/test/test_reconnect.py]: Yes (for a short time)

File: stp_zmq/test/test_reconnect.py (hunk 3 above)
Comment (ashcherbakov) [inline stp_zmq/test/test_reconnect.py]: Removed duplicate test

File: stp_zmq/test/test_reconnect.py (hunk 4 above)
Comment (ashcherbakov) [inline stp_zmq/test/test_reconnect.py]: Removed duplicate test

File: stp_zmq/test/test_reconnect.py (hunk 5 above)
Comment (ashcherbakov) [inline stp_zmq/test/test_reconnect.py]: Moved to a function

**Pull request merged**